### PR TITLE
Changed processing flow of invoices to handle one invoice at a time. …

### PR DESCRIPTION
**Pull request by AndreSoderlund**
…Added database support for storing the items and their statuses for troubleshooting. This database support also gives us the possibility to re-run batchesin the future

## Types of changes

- [ ] Bug fix
- [x] New feature
- [ ] Removed feature
- [x] Code style update (formatting etc.)
- [x] Refactoring (no functional changes, no api changes)
- [ ] Build related changes
- [ ] Documentation content changes
- [ ] Content/Data

## Does this PR introduce a breaking change?
- [ ] Yes (I have stepped the version number accordingly)
- [x] No
      
## Checklist:

- [x] My code follows the code style of this project.
- [x] I have updated the documentation accordingly (if applicable).
- [x] I have added/updated tests to cover my changes (if applicable).

### CODE REVIEW

File: src/integration-test/java/apptest/InvoiceSenderIT.java [hunk 1]
```diff
@@ -1,109 +1,216 @@
...
-import java.time.Duration;
+import jcifs.CIFSContext;
+import jcifs.smb.SmbFile;
 import org.apache.commons.io.FileUtils;
-import org.junit.jupiter.api.AfterEach;
-import org.junit.jupiter.api.BeforeEach;
-import org.junit.jupiter.api.Disabled;
+import org.junit.jupiter.api.AfterAll;
 import org.junit.jupiter.api.Test;
+import org.springframework.beans.factory.annotation.Autowired;
```
Comment (cheezi0747) [inline src/integration-test/java/apptest/InvoiceSenderIT.java:8]: Just FYI. Your import settings are not up to date. use [this](https://github.com/Sundsvallskommun/.github-private/blob/main/profile/development-guidelines-api-team/sundsvall_profile_idea.xml) or you will have headaches when the next dept44 version rolls out

File: src/main/java/se/sundsvall/invoicesender/integration/db/entity/ItemEntity.java [hunk 2]
```diff
@@ -40,6 +57,71 @@ public void setId(final Integer id) {
...
+
+	public void setType(ItemType type) {
+		this.type = type;
+	}
+
+	public ItemEntity withType(ItemType type) {
+		this.type = type;
+		return this;
+	}
+
```
Comment (cheezi0747) [inline src/main/java/se/sundsvall/invoicesender/integration/db/entity/ItemEntity.java:124]: No ToString / Equals / Haschcode?

File: src/main/java/se/sundsvall/invoicesender/service/model/Metadata.java [hunk 3]
```diff
@@ -0,0 +1,118 @@
...
+
+	public Metadata withReminder(final boolean reminder) {
+		this.reminder = reminder;
+		return this;
+	}
+
+	public void setReminder(final boolean reminder) {
+		this.reminder = reminder;
+	}
+
```
Comment (cheezi0747) [inline src/main/java/se/sundsvall/invoicesender/service/model/Metadata.java:117]: toString / Equals / hashcode

File: src/integration-test/resources/InvoiceSenderIT/__files/test3_processInvoices/mappings/api-citizen-get-person.json [hunk 4]
```diff
@@ -0,0 +1,11 @@
+{
+	"request": {
+		"method": "GET",
+		"urlPattern": "/citizen/[0-9a-fA-F]{8}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{4}-[0-9a-fA-F]{12}"
+	},
+	"response": {
+		"headers": {"Content-Type": "application/json"},
+		"status": 200
```
Comment (josc04) [inline src/integration-test/resources/InvoiceSenderIT/__files/test3_processInvoices/mappings/api-citizen-get-person.json:8]: No response body?

File: src/main/java/se/sundsvall/invoicesender/integration/db/entity/ItemType.java [hunk 5]
```diff
@@ -0,0 +1,7 @@
+package se.sundsvall.invoicesender.integration.db.entity;
+
+public enum ItemType {
+	UNKNOWN,
```
Comment (josc04) [inline src/main/java/se/sundsvall/invoicesender/integration/db/entity/ItemType.java:4]: Out of curiosity, what the difference between UNKNOWN and OTHER?

File: src/main/resources/db/migration/V1.7__add_columns.sql
Comment (josc04) [inline src/main/resources/db/migration/V1.7__add_columns.sql:1]: No script for modifying existing entries regarding 
- type (checked production db and it seems to only contain entries of type INVOICE)
- processing_enabled - I do not know how this is used, but by default all entries will set this to 1 now, is that correct?
- completed - same here, shouldn't existing entries have this set to 1 depending on their status (NOT_SENT, SENT, IGNORED etc)?

File: src/test/java/se/sundsvall/invoicesender/integration/db/DbIntegrationTests.java [hunk 6]
```diff
@@ -55,66 +51,42 @@ void testGetBatches() {
 		assertThat(result.getNumber()).isZero();
 		assertThat(result.getTotalPages()).isEqualTo(2);
 
-		verify(mockBatchRepository, times(1)).findAllByCompletedAtBetweenAndMunicipalityId(
+		verify(batchRepositoryMock, times(1)).findAllByCompletedAtBetweenAndMunicipalityId(
```
Comment (josc04) [inline src/test/java/se/sundsvall/invoicesender/integration/db/DbIntegrationTests.java:54]: `times(1)` is redundant

File: src/main/java/se/sundsvall/invoicesender/integration/db/entity/ItemEntity.java (hunk 2 above)
Comment (AndreSoderlund) [reply src/main/java/se/sundsvall/invoicesender/integration/db/entity/ItemEntity.java:124]: None of them are used in this application.

File: src/integration-test/resources/InvoiceSenderIT/__files/test3_processInvoices/mappings/api-citizen-get-person.json (hunk 4 above)
Comment (AndreSoderlund) [reply src/integration-test/resources/InvoiceSenderIT/__files/test3_processInvoices/mappings/api-citizen-get-person.json:8]: No, we are using this just to check for status code. If a person have protected identity, the response code will not be 200. 

```
@GetMapping(path = "/{personId}", produces = ALL_VALUE)
ResponseEntity<Void> getPerson(@PathVariable("personId") String personId);
```

File: src/main/java/se/sundsvall/invoicesender/integration/db/entity/ItemType.java (hunk 5 above)
Comment (AndreSoderlund) [reply src/main/java/se/sundsvall/invoicesender/integration/db/entity/ItemType.java:4]: When an item is first persisted, the type is unknown. 

At a later stage we check the type for each item. In that check we either change the type to INVOICE or OTHER. At this stage, the type is known and hence no longer unknown.

**Review by cheezi0747 — DISMISSED** [review]

File: src/main/java/se/sundsvall/invoicesender/integration/raindance/RaindanceIntegration.java [hunk 7]
```diff
@@ -31,22 +37,15 @@
...
+import se.sundsvall.invoicesender.integration.db.entity.ItemEntity;
 
 public class RaindanceIntegration {
 
 	private static final Logger LOG = LoggerFactory.getLogger(RaindanceIntegration.class);
 
 	private static final DateTimeFormatter DATE_FORMATTER = DateTimeFormatter.ofPattern("yyMMdd");
-	private static final Predicate<Item> UNSENT_ITEMS = item -> item.getStatus() != SENT;
+	private static final Predicate<ItemEntity> UNSENT_ITEMS = item -> item.getStatus() != SENT;
 	private static final String BATCH_FILE_SUFFIX = ".zip.7z";
```
Comment (jklingst) [inline src/main/java/se/sundsvall/invoicesender/integration/raindance/RaindanceIntegration.java:49]: You could use the `BATCH_FILE_SUFFIX` constant from the `Constants` class instead.

File: src/main/java/se/sundsvall/invoicesender/util/Constants.java [hunk 8]
```diff
@@ -0,0 +1,35 @@
...
+import java.util.regex.Pattern;
+import se.sundsvall.invoicesender.integration.db.entity.ItemEntity;
+
+public final class Constants {
+
+	private Constants() {
+
+	}
+
+	public static final Predicate<ItemEntity> ITEM_IS_A_PDF = item -> item.getFilename().toLowerCase().endsWith(".pdf");
```
Comment (jklingst) [inline src/main/java/se/sundsvall/invoicesender/util/Constants.java:21]: Personal taste, but I would have left these predicates in the `ItemEntity` class

**Review by jklingst — APPROVED** [review]

**Pull request merged**